ibis/expr/types/core.py

```diff
@@ -219,13 +223,34 @@ def _find_backends(self) -> list[BaseBackend]:
                     elif isinstance(arg, Expr):
                         stack.append(arg.op())
 
-        return list(seen_backends.values())
+        return list(seen_backends.values()), has_unbound
 
-    def _find_backend(self) -> BaseBackend:
-        backends = self._find_backends()
+    def _find_backend(self, *, use_default: bool = False) -> BaseBackend:
+        """Find the backend attached to an expression.
+
+        Parameters
+        ----------
+        use_default
+            If [`True`][True] and the default backend isn't set, initialize the
+            default backend and use that. This should only be set to `True` for
+            `.execute()`. For other contexts such as compilation, this option
+            doesn't make sense so the default value is [`False`][False].
+
+        Returns
+        -------
+        BaseBackend
+            A backend that is attached to the expression
+        """
+        backends, has_unbound = self._find_backends()
 
         if not backends:
-            default = config.options.default_backend
+            if has_unbound:
+                raise IbisError(
+                    "Expression contains unbound tables and therefore cannot "
```

Comment:
We could be a bit more verbose here and tell the user to `ibis.<backend>.execute()` or set a default backend.

Reply:
Also, could you please cover it with a test case so we ensure that this error reporting remains persistent after the refactors?

Reply:
Yep!

Reply:
Done!